hermes/probe/metric.go

```diff
@@ -0,0 +1,107 @@
+type probeRunMetrics struct {
+	StorageSystem string
+	Target        string
+	BucketName    string
+
+	// probeOpLatency is used to record latency values with distinct labels
+	// per exit status per probe operation per target.
+	// Recommended usage: probeOpLatency[probeOp][exitStatus].Metric("latency").AddFloat64(<val>)
+	ProbeOpLatency map[string]map[string]*metrics.EventMetrics
+	// apiCallLatency is used to record latency values with distinct labels
+	// per exit status per API call per target.
+	// Recommended usage: apiCallLatency[apiCall][exitStatus].Metric("latency").AddFloat64(<val>)
+	ApiCallLatency map[string]map[string]*metrics.EventMetrics
+}
+
+// NewProbeMetrics creates a new *probeRunMetrics object and initialises the fields inside it.
+// Arguments:
+//	- conf: pass a HermesProbeDef config
+//	- target: pass the target for which metrics are to be collected.
+// Returns:
+//	- pm: returns an initialised *probeRunMetrics object.
+//	- err: returns an error if a latency distribution cannot be created from the config proto.
+func NewProbeMetrics(conf probepb.HermesProbeDef, target probepb.Target) (*probeRunMetrics, error) {
+	probeOps := []string{"total_probe_run", "check_nil", "read_and_verify", "delete_file", "create_file"}
+	apiCalls := []string{"list_files", "create_file", "delete_file", "get_file"}
+	probeExitStatuses := []string{"success", "op_timeout", "probe_failed", "file_missing", "bucket_missing", "file_corrupted", "file_read_failure", "file_metadata_mismatch", "unknown_file_found", "all_files_missing"}
+	apiExitStatuses := []string{"success", "api_call_failed", "file_missing", "bucket_missing", "file_corrupted", "file_read_failure", "file_metadata_mismatch", "unknown_file_found", "all_files_missing"}
+
+	pm := &probeRunMetrics{
+		StorageSystem: target.GetTargetSystem().String(),
+		Target:        target.GetName(),
+		BucketName:    target.GetBucketName(),
+	}
+	var err error
+	if pm.ProbeOpLatency, err = pm.initLatencyDistMetric(conf.GetProbeLatencyDistribution(), probeOps, probeExitStatuses); err != nil {
+		return nil, err
+	}
+
+	if pm.ApiCallLatency, err = pm.initLatencyDistMetric(conf.GetApiCallLatencyDistribution(), apiCalls, apiExitStatuses); err != nil {
+		return nil, err
+	}
+
+	return pm, nil
+}
+
+// initLatencyDistMetric initialises a latency distribution metric with the given arguments.
+// Arguments:
+//	- latDistPb: pass the proto definition of a distribution metric.
+//	- ops: pass the list of operations associated with this metric.
+//	- exitStatuses: pass the list of possible exit statuses associated with this metric.
+// Returns:
+//	- latencyMetric: returns an initialises latency metric map.
+//	- err: returns an error if there a problem creating a distribution metric from the proto passed.
+func (p *probeRunMetrics) initLatencyDistMetric(latDistPb *metricpb.Dist, ops, exitStatuses []string) (map[string]map[string]*metrics.EventMetrics, error) {
```

Comment:
uber nit: to make naming consistent with that of NewProbeMetrics, consider renaming `p` to `pm`.
```suggestion
func (pm *probeRunMetrics) initLatencyDistMetric(latDistPb *metricpb.Dist, ops, exitStatuses []string) (map[string]map[string]*metrics.EventMetrics, error) {
```

Reply:
I've renamed this to `Metrics` and used `m`, which satisfies this too :)